File: androidApp/src/main/java/com/example/stories/android/util/ui/LoadingDataScreen.kt
package com.example.stories.android.util.ui

import android.widget.Toast
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonColors
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.min
import com.example.stories.SharedRes
import com.example.stories.android.ui.StoriesTheme
import com.example.stories.android.util.resources.getString
import com.example.stories.android.util.resources.getStringResource
import com.example.stories.android.util.resources.getPainterResource
import com.example.stories.infrastructure.loading.LoadingError
import com.example.stories.infrastructure.loading.LoadStatus
import dev.icerock.moko.resources.ImageResource

@Composable
fun<T: Any> RefreshLoadingDataScreen(
    loadStatus: LoadStatus<T>,
    onRefresh: (showLoading: Boolean) -> Unit,
    isDataEmpty: (T) -> Boolean = { false },
    refreshTitle: String = getStringResource { empty_generic_screen_title },
    errorContent: @Composable (LoadingError) -> Unit = {
        DefaultErrorScreen(
            loadingError = it,
            buttonText = getStringResource { refresh },
            onClickButton = { onRefresh(true) },
        )
    },
    loadingContent: @Composable () -> Unit = { DefaultLoadingScreen() },
    successContent: @Composable BoxScope.(T) -> Unit,
) {
    loadStatus.fold(
        onError = { errorContent(it) },
        onLoading = { loadingContent() },
        onSuccess = {
            if (isDataEmpty(it)) {
                EmptyDataScreen(
                    title = refreshTitle,
                    onClickButton = { onRefresh(true) },
                )
            } else {
                PullRefreshLayout(
                    isRefreshing = loadStatus.isRefreshing(),
                    onRefresh = { onRefresh(false) },
                ) {
                    successContent(it)
                }
            }
        },
    )
}

@Composable
fun<T: Any> LoadingDataScreen(
    loadStatus: LoadStatus<T>,
    errorContent: @Composable (LoadingError) -> Unit = { DefaultErrorScreen(it) },
    loadingContent: @Composable () -> Unit = { DefaultLoadingScreen() },
    successContent: @Composable (T) -> Unit
) {
    loadStatus.fold(
        onError = { errorContent(it) },
        onLoading = { loadingContent() },
        onSuccess = { successContent(it) },
    )
}

@Composable
fun InfoScreen(
    icon: ImageResource,
    title: String,
    message: String?,
    buttonText: String = getStringResource { accept },
    colors: ButtonColors = ButtonDefaults.buttonColors(),
    onClickButton: (() -> Unit)? = null,
) {
    Box(
        modifier = Modifier
            .fillMaxSize()
            .padding(24f.dp)
    ) {
        Column(
            modifier = Modifier.fillMaxSize(),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center
        ) {
            Image(
                painter = getPainterResource(imageResource = icon),
                modifier = Modifier
                    .padding(16.dp)
                    .size(36f.dp),
                contentDescription = "",
            )
            Text(
                text = title,
                textAlign = TextAlign.Center,
                style = MaterialTheme.typography.headlineMedium,
            )
            Text(
                text = message ?: return@Column,
                textAlign = TextAlign.Center,
                modifier = Modifier
                    .padding(top = 30f.dp)
            )
        }
        Button(
            onClick = onClickButton ?: return,
            modifier = Modifier
                .align(alignment = Alignment.BottomCenter)
                .fillMaxWidth(),
            colors = colors
        ) {
            Text(text = buttonText)
        }
    }
}

@Composable
fun DefaultErrorScreen(
    loadingError: LoadingError,
    buttonText: String = getStringResource { accept },
    onClickButton: (() -> Unit)? = null,
) {
    InfoScreen(
        icon = loadingError.icon,
        title = loadingError.title.getString(),
        message = loadingError.message?.getString(),
        buttonText = buttonText,
        colors = ButtonDefaults.buttonColors(
            containerColor = MaterialTheme.colorScheme.error,
            contentColor = MaterialTheme.colorScheme.onError
        ),
        onClickButton = onClickButton,
    )
}

@Preview
@Composable
fun DefaultErrorScreen_preview() {
    StoriesTheme {
        Surface(
            color = MaterialTheme.colorScheme.background
        ) {
            val context = LocalContext.current
            val toast = remember {
                Toast.makeText(
                    context,
                    "Can do nothing bro, this is just a preview",
                    Toast.LENGTH_LONG
                )
            }
            DefaultErrorScreen(
                loadingError = LoadingError.GenericError,
                onClickButton = toast::show
            )
        }
    }
}

@Composable
fun EmptyDataScreen(
    title: String = getStringResource { empty_generic_screen_title },
    onClickButton: () -> Unit
) {
    InfoScreen(
        icon = SharedRes.images.empty_list,
        title = title,
        message = getStringResource { empty_screen_description },
        buttonText = getStringResource { refresh },
        onClickButton = onClickButton
    )
}

@Preview
@Composable
fun EmptyDataScreen_preview() {
    StoriesTheme {
        Surface(
            color = MaterialTheme.colorScheme.background
        ) {
            EmptyDataScreen(onClickButton = {})
        }
    }
}

@Composable
fun DefaultLoadingScreen() {
    BoxWithConstraints(
        modifier = Modifier.fillMaxSize()
    ) {
        val dimen = min(maxHeight, maxWidth) / 3
        CircularProgressIndicator(
            modifier = Modifier
                .size(dimen)
                .align(Alignment.Center),
            strokeCap = StrokeCap.Round
        )
    }
}

@Preview
@Composable
fun DefaultLoadingScreen_preview() {
    StoriesTheme {
        Surface(
            color = MaterialTheme.colorScheme.background
        ) {
            DefaultLoadingScreen()
        }
    }
}

File: iosApp/iosApp/util/ui/LoadingDataScreen.swift
import SwiftUI
import shared

struct RefreshLoadingDataScreen<Data: AnyObject, Content, LoadingContent>: View where Content : View, LoadingContent : View {
    
    let loadStatus: LoadStatus<Data>?
    let onRefresh: (Bool) -> Void
    let isDataEmpty: (Data) -> Bool
    let refreshTitle: String
    @ViewBuilder let errorContent: (LoadingError) -> any View
    @ViewBuilder let loadingContent: () -> LoadingContent
    @ViewBuilder let successContent: (Data) -> Content
    
    init(
        loadStatus: LoadStatus<Data>?,
        onRefresh: @escaping (Bool) -> Void,
        isDataEmpty: @escaping (Data) -> Bool = { _ in false},
        refreshTitle: String = getStringResource(path: \.empty_generic_screen_title),
        @ViewBuilder errorContent: @escaping (LoadingError) -> any View,
        @ViewBuilder loadingContent: @escaping () -> LoadingContent = {
            DefaultLoadingScreen()
        },
        @ViewBuilder successContent: @escaping (Data) -> Content
    ) {
        self.init(
            loadStatus: loadStatus,
            onRefresh: onRefresh,
            isDataEmpty: isDataEmpty,
            refreshTitle: refreshTitle,
            optionalErrorContent: errorContent,
            loadingContent: loadingContent,
            successContent: successContent
        )
    }
    
    init(
        loadStatus: LoadStatus<Data>?,
        onRefresh: @escaping (Bool) -> Void,
        isDataEmpty: @escaping (Data) -> Bool = { _ in false},
        refreshTitle: String = getStringResource(path: \.empty_generic_screen_title),
        @ViewBuilder loadingContent: @escaping () -> LoadingContent = { DefaultLoadingScreen() },
        @ViewBuilder successContent: @escaping (Data) -> Content
    ) {
        self.init(
            loadStatus: loadStatus,
            onRefresh: onRefresh,
            isDataEmpty: isDataEmpty,
            refreshTitle: refreshTitle,
            optionalErrorContent: nil,
            loadingContent: loadingContent,
            successContent: successContent
        )
    }
    
    private init(
        loadStatus: LoadStatus<Data>?,
        onRefresh: @escaping (Bool) -> Void,
        isDataEmpty: @escaping (Data) -> Bool,
        refreshTitle: String,
        optionalErrorContent: ((LoadingError) -> any View)? = nil,
        @ViewBuilder loadingContent: @escaping () -> LoadingContent = { DefaultLoadingScreen() },
        @ViewBuilder successContent: @escaping (Data) -> Content
    ) {
        self.loadStatus = loadStatus
        self.onRefresh = onRefresh
        self.isDataEmpty = isDataEmpty
        self.refreshTitle = refreshTitle
        self.errorContent = optionalErrorContent ?? { error in
            DefaultErrorScreen(
                loadingError: error,
                onClickEnabled: true,
                onClickButton: { onRefresh(true) },
                buttonText: getStringResource(path: \.refresh)
            )
        }
        self.loadingContent = loadingContent
        self.successContent = successContent
    }
    
    var body: some View {
        if let data = loadStatus?.dataOrNull() {
            if isDataEmpty(data) {
                EmptyDataScreen(
                    title: refreshTitle,
                    onClickButton: { onRefresh(true) }
                )
            } else {
                let isRefreshing = loadStatus?.isRefreshing() == true
                PullRefreshLayout(
                    isRefreshing: isRefreshing,
                    onRefresh: { onRefresh(false) }
                ) {
                    successContent(data)
                }
            }
        } else if let error = loadStatus?.errorOrNull() {
            AnyView(errorContent(error))
        } else {
            loadingContent()
        }
    }
}

struct LoadingDataScreen<Data: AnyObject, Content, ErrorContent, LoadingContent>: View where Content : View, ErrorContent : View, LoadingContent : View {
    
    let loadStatus: LoadStatus<Data>?
    @ViewBuilder let errorContent: (LoadingError) -> ErrorContent
    @ViewBuilder let loadingContent: () -> LoadingContent
    @ViewBuilder let successContent: (Data) -> Content
    
    init(
        loadStatus: LoadStatus<Data>?,
        @ViewBuilder errorContent: @escaping (LoadingError) -> ErrorContent = { error in
            DefaultErrorScreen(loadingError: error, onClickEnabled: false, onClickButton: { })
        },
        @ViewBuilder loadingContent: @escaping () -> LoadingContent = { DefaultLoadingScreen() },
        @ViewBuilder successContent: @escaping (Data) -> Content
    ) {
        self.loadStatus = loadStatus
        self.errorContent = errorContent
        self.loadingContent = loadingContent
        self.successContent = successContent
    }
    
    var body: some View {
        if let data = loadStatus?.dataOrNull() {
            successContent(data)
        } else if let error = loadStatus?.errorOrNull() {
            errorContent(error)
        } else {
            loadingContent()
        }
    }
}

struct LoadingDataScreen_Previews: PreviewProvider {
    static var previews: some View {
        LoadingDataScreen(
            loadStatus: HistoryMocks().getMockLoadStatusStrings()
        ) { error in
            DefaultErrorScreen(loadingError: error, onClickEnabled: false, onClickButton: { })
        } loadingContent: {
            DefaultLoadingScreen()
        } successContent: { data in
            if let texts = data as? [String] {
                VStack {
                    Text(texts[0])
                    Text(texts[1])
                    Text(texts[2])
                }
            }
        }
    }
}

struct InfoScreen : View {
    let icon: shared.ImageResource
    let title: String
    let message: String?
    let onClickEnabled: Bool
    let onClickButton: () -> Void
    let buttonText: String
    let buttonModifier: CustomModifier

    init(
        icon: shared.ImageResource,
        title: String,
        message: String?,
        onClickEnabled: Bool,
        onClickButton: @escaping () -> Void,
        buttonText: String? = nil,
        buttonModifier: CustomModifier? = nil
    ) {
        self.icon = icon
        self.title = title
        self.message = message
        self.onClickEnabled = onClickEnabled
        self.onClickButton = onClickButton
        self.buttonText = buttonText ?? getStringResource(path: \.accept)
        self.buttonModifier = buttonModifier ?? CustomModifier()
    }

    var body: some View {
        ZStack {
            VStack {
                Image(uiImage: icon.toUIImage()!)
                    .resizable()
                    .frame(width: 36.0, height: 36.0)
                    .padding()
                
                Text(title)
                    .multilineTextAlignment(.center)
                    .font(.largeTitle)
                
                if let message = message {
                    Text(message).padding()
                        .multilineTextAlignment(.center)
                }
            }
            
            if onClickEnabled {
                VStack {
                    Spacer()
                    Button(action: onClickButton) {
                        Text(buttonText)
                            .padding(6)
                            .frame(maxWidth: .infinity)
                            .modifier(buttonModifier)
                            .clipShape(RoundedRectangle(cornerRadius: 4))
                    }
                }
            }
        }.padding()
    }
}

struct DefaultErrorScreen : View {
    
    let loadingError: LoadingError
    let onClickEnabled: Bool
    let onClickButton: () -> Void
    let buttonText: String

    init(
        loadingError: LoadingError,
        onClickEnabled: Bool,
        onClickButton: @escaping () -> Void,
        buttonText: String? = nil
    ) {
        self.loadingError = loadingError
        self.onClickEnabled = onClickEnabled
        self.onClickButton = onClickButton
        self.buttonText = buttonText ?? getStringResource(path: \.accept)
    }

    var body: some View {
        InfoScreen(
            icon: loadingError.icon,
            title: loadingError.title.getString(),
            message: loadingError.message?.getString(),
            onClickEnabled: onClickEnabled,
            onClickButton: onClickButton,
            buttonText: buttonText,
            buttonModifier: CustomModifier {
                $0.background(Color(getColorResource(path: \.error)))
                    .foregroundColor(Color.black)
            }
        )
    }
}

struct DefaultErrorScreen_Previews: PreviewProvider {
    static var previews: some View {
        DefaultErrorScreen(
            loadingError: LoadingError.GenericError(),
            onClickEnabled: true,
            onClickButton: { }
        )
    }
}

struct EmptyDataScreen : View {
    let title: String
    let onClickButton: () -> Void
    
    init(
        title: String = getStringResource(path: \.empty_generic_screen_title),
        onClickButton: @escaping () -> Void
    ) {
        self.title = title
        self.onClickButton = onClickButton
    }
    
    var body: some View {
        InfoScreen(
            icon: SharedRes.images().empty_list,
            title: title,
            message: getStringResource(path: \.empty_screen_description),
            onClickEnabled: true,
            onClickButton: onClickButton,
            buttonText: getStringResource(path: \.refresh),
            buttonModifier: CustomModifier()
        )
    }
}

struct EmptyDataScreen_preview : PreviewProvider {
    static var previews: some View {
        EmptyDataScreen {}
    }
}


struct DefaultLoadingScreen : View {
        
    var body: some View {
        ProgressView()
            .controlSize(.large)
    }
}

struct DefaultLoadingScreen_Previews: PreviewProvider {
    static var previews: some View {
        DefaultLoadingScreen()
    }
}
